### ISSUE TITLE
Test nk3 related hotp-verification patches upstream

### ISSUE THREAD

**Issue by tlaurion** [linuxboot/heads#1866]
Focusing on PR content, see PR to follow white rabbit on security/UX/oem [issues](https://github.com/Nitrokey/nitrokey-hotp-verification/issues/36) they solve:

- https://github.com/Nitrokey/nitrokey-hotp-verification/pull/43
  - attempts to resolve issue https://github.com/Nitrokey/nitrokey-hotp-verification/issues/38
- https://github.com/Nitrokey/nitrokey-hotp-verification/pull/44
  - attempts to resolve issue https://github.com/Nitrokey/nitrokey-hotp-verification/issues/39
- https://github.com/Nitrokey/nitrokey-hotp-verification/pull/45
  - unneeded since hotp_verification reset doesn't set a secure app PIN for now...
- https://github.com/Nitrokey/nitrokey-hotp-verification/pull/46
  - attempts to solve https://github.com/Nitrokey/nitrokey-hotp-verification/issues/42 by requiring hotp_verification reset to implement an optional PIN argument since not setting one by default, preventing PIN change...

- hotp_verification version bump plus upstream commits not part of 1.7 under https://github.com/Nitrokey/nitrokey-hotp-verification/issues/52 to suppress bogus touch related output still not being clean
 
40h of work and counting. Will edit. 

---

2024-12-16: Heads-->Nitrokey work highlighted above still refused to be paid in totality without fighting as of now. 
Will have call with @jans23 Tuesday 2024-12-17 after mail exchanged but dismissed, to see if we can arrive to respectful collaboration or not for the future. 

In all case, in absence of co-beneficial agreement (mutualism), I will not test Nitrokey code unless it's associated to a PR under Heads from now on since Heads related work should be paid. I'm responsible for Heads, and time must be recognized/valued/paid accordingly **Heads partnership should be based on biological concept of mutualism and nothing else**. This work, in this issue, should not be confounded with what should be expected to be paid by profit sharing agreement (maintenance and new board porting collaboration related only). If i'm expected to fix Nitrokey stuff, then Nitrokey must pay for the past, present and in the future.

_If no mutualism is possible with Nitrokey, at bare minimum commensalism would be expected. 
But in no case I will tolerate parasitism from Nitrokey._ 

Heads usage of Nitrokey based devices started by mutialism with Purism partnering with Heads in 2018 to ease remote attestation through reverse HOTP secret sealing on USB security dongles of their rebranded Nitrokey 1 pro devices, details at https://www.linuxjournal.com/content/tamper-evident-boot-heads. Mutualism could happen again if Nitrokey recognizes their responsibility in making tamper evidence a priority again, but there is not much more of my free time I can invest into this nk3 where absolutely no collaboration from Nitrokey toward Heads (Nitrokey->Heads participation in code/money) and Where Heads invested too much time (Heads->Nitrokey) for this to be considered anything else than parasitism from my perspective at this point.

If not comfortable about those biology principles, make your head reading https://biodifferences.com/difference-between-mutualism-commensalism-and-parasitism.html

----

@jans23 donations expected otherwise expect less collaboration from me in the future, and future security issues discovered on my side will be associated with a CVE. I learned my lessons, and won't hesitate on going public following ethical disclosure, which i did. This was discussed in chat, in person, you don't seem to understand.

This is me reminding you that the nk3 is not a novelty product and used for security purposes, here under Heads for remote attestation. This remote attestation, through reverse HOTP to a usb security dongle under Heads, always required: authentication, reset and provisioning capabilities, and since forever forwrd plans discussed many times in the past to reduce OEM provisioning time, unattended reset+automated secret provisioning to augment strength of anti-interdiction (both hardware+security dongle in same hipping package) while reducing OEM provisioning costs: reduce time to shipment on OEM side and ease User Re-Ownership frictions while implementing best practices with even diceware passphrases (I shipped this with the Privacy Beast at first shipment back in ~2018, made conferenes about this, and that was what was QubesOS certified.)

If work leading to https://www.nitrokey.com/blog/2024/heads-v25-and-nitrokey-3-firmware-v171-security-update was not compensated in money before (+30h there, didn't count), and as we talked at QubesOS mini-summit 2024 in person, work needs to be paid, otherwise maybe stop selling security promises or pay for security audits not relying on my discoveries that shake trust from the ecosystem and myself into secirity promises, discoveries from my side while testing other things to discover vulnerabilities/bad design/regressions in nk3 over previous security dongles. This impacts myself, Nitrokey, but also partners and the whole ecosystem, resulting in money losses from every support team depending on your security dongles, resulting in issues opened in all partners issue tracking systems, qubesos forum discussions and ping pong games between everyone; meaning downstream/upstream lost time and energies resulting in money loss by being neglectful. 

The issues/PR above, and as you can see if clicking on them, needed a lot of involvement (again) ffrom my side, which was not expected/planned. I understand that it was not planned from Nitrokey either, but that is the problem. And work needs to be paid. I never consented to this in this feature freeze, nor what lead to 1.7.1 nk3 firmware when testing improvements on OEM factory reset/Re-Ownership 6 months ago either.

Details of work done here:
- nk3 reset needed to done by Heads as part of oem-factory reset and re-onwership. Still requires physical presence which is unexplained regression of nk3 and prevents unattended oem reownership with diceware secrets for now until regression fix for https://github.com/Nitrokey/nitrokey-hotp-verification/issues/41
- nk3 needed to provide real information on its secrets app PIN state, which is not the same as for nk2/nk1/nk2 storage (which was GPG Admin PIN) to be parsed and dealt with so that sealing HOTP could be done when public key <1 month (in delay of oem factory reset to the moment shipped to the user on first HOTP sealing per OEM/User attempting default PIN). This was broken since nk3 release and never fixed by Nitrokey nor under Heads code or even discussed prior of introducing change and regression, leading to losses.
- Heads needed to be adapted to deal with proper hotp-verification output for nk3 now reported secrets app PIN (https://github.com/Nitrokey/nitrokey-hotp-verification/pull/43)

This can be tested under https://github.com/linuxboot/heads/pull/1875 and I won't have time to split it all out, feature freeze discussed was 2024-11-20, and as of 2024-12-11 I'm still waiting for a version bump of hotp-verification to use that commit under Heads.

**Comment by tlaurion** [comment]
The one i'm the most interested (Heads maintainer) is https://github.com/Nitrokey/nitrokey-hotp-verification/pull/46, starting with it.

**Comment by tlaurion** [comment]
> The one i'm the most interested (Heads maintainer) is [Nitrokey/nitrokey-hotp-verification#46](https://github.com/Nitrokey/nitrokey-hotp-verification/pull/46), starting with it.

Ha. segfaults on nk2/lk https://github.com/Nitrokey/nitrokey-hotp-verification/pull/46#issuecomment-2503442845

**Comment by tlaurion** [comment]
Finally got an understanding that it doesn't make sense to not set a pin if no default pin is set at https://github.com/Nitrokey/nitrokey-hotp-verification/pull/46#issuecomment-2508143760

**Comment by tlaurion** [comment]
There is still two pins instead of one at https://github.com/Nitrokey/nitrokey-hotp-verification/pull/44#issuecomment-2508687329

**Comment by tlaurion** [comment]
A lot of misunderstanding around https://github.com/Nitrokey/nitrokey-hotp-verification/pull/45 around related issues.

There is no need to change pins if there is only pin that if locked requires reset, as opposed to gpg pins.....


Under gpg:
- reset code unlocks gpg Admin pin
- gpg ad in pin resets lock of user pin
- user pin cannot unlock itself.


On devices prior of nk3
- hotp related pin was Admin pin
- Admin PIN was resettled to default pin 12345678 per re-ownership
- Admin PIN and user pin coukd be changed because the default PINs were known

Nk3:
- adds a secret app pin
- currently unset by default
- pin set in first use since hotp 1.6 and nk3 firmware 1.7.1. Prior, was accepting any pin, hotp secret resealing was only needing physical presence



Therefore.
- It makes no sense to have a pin change if no default pin is set
- secure element pin has a counter of 6 attempts as opposed to 3 for gpg Admin PIN on devices <nk3
- changing pin use case would be if user upgraded firmware and fears eavesdropping of that pin because.... He upgraded firmware in unsafe environement? Bad user.


TLDR...... 
> - Either set a default secret app pin and offer pin change so re-ownership can change it just like before as part of re-ownership for <nk3
- have `hotp_verification reset SECRET_APP_PIN` requiring a pin
- Have `hotp_verification reset` set a default PIN, which if we don't plan to reinvent the wheel should be equivalent to gpg Admin PIN which is 12345678.

@daringer this is heads requirements. You have to decide what is best for nitrokey, but this is looping. You choose to reinvent the wheel without consulting first. And current implementation is bogus.

**Comment by tlaurion** [comment]
We seem to finally have an understanding that
- hotp_verification info alone is needed, there is no need for pin-info, and that info should present only real, accurate, actionnable info: GPG Admin/User PIN, Secure App PIN, fimware version (user can check it: 1.7.1? no need for secure app subversion!), and this info being output only if present (no Secure App PIN: don't show it) under https://github.com/Nitrokey/nitrokey-hotp-verification/pull/44
- We seem to finally have agreed that the way to go for `hotp_verification reset` **will not be to set a default PIN by default**, _and require SECRET_APP_PIN argument to be provided by oem-factory-reset under Heads_, which will require physical presence until next nk3 version. This is ongoing https://github.com/Nitrokey/nitrokey-hotp-verification/pull/46


Therefore:
- secure app PIN passwd change is irrelevant if no PIN is set by default, and would require 2x physical presence (touch).
- pin-info is irrelevant (info is relevant) with requirements above, where we concluded that an additional subset of Admin/User PIN (which doesn't exist: there is only a single Secure App PIN existing)

Please confirm that we have same understanding @sosthene-nitrokey @daringer @nestire and this actionaable where I can only plug in change of commits and go forward in this issue which is aimed to test until mergeable and not causing regression for <nk3 and makes sense for >nk3.

I implemented a workaround for now for gpg output of PINs prior of asking to unlock openpgp smartcard for signing (showing GPG User PIN rety count left before User PIN locks see latest video output https://github.com/linuxboot/heads/pull/1863), and reduce support request volumes, and will apply proper safeguards under Heads as started under https://github.com/linuxboot/heads/pull/1850.

**Comment by tlaurion** [comment]
Another 5 hours today, customizing messages on Heads side and testing reset APP_SECURE_APP PIN in oem-factory reset.

Adding 5 hours to the 15 in OP.
https://github.com/linuxboot/heads/compare/b550151d54fb8545b16ff4e6c9a2e2eb57f0d289..444ff3ee3719956744f780e558bb3d72890e1224

**Comment by tlaurion** [comment]
5 hours testing https://github.com/Nitrokey/nitrokey-hotp-verification/pull/43 https://github.com/Nitrokey/nitrokey-hotp-verification/pull/46 under https://github.com/linuxboot/heads/pull/1850

Modifying OP from 20h hours of work to 25h

**Comment by tlaurion** [comment]
5 hours https://github.com/Nitrokey/nitrokey-hotp-verification/pull/46#issuecomment-2528913089

Changing op from 25h to 30h

**Comment by tlaurion** [comment]
hopt-verification still not version bumped. Not doing anything. Waiting. Seems like this work won't be paid.

**Comment by tlaurion** [comment]
Version bump testing under heads, plus testing of some commits upstreamed not part of 1.7 resulting of issue https://github.com/Nitrokey/nitrokey-hotp-verification/issues/52 and other issues/PR. 

10h more, modifying op to 40h

**Comment by tlaurion** [comment]
https://github.com/Nitrokey/nitrokey-hotp-verification/pull/51  Nitrokey/nitrokey-hotp-verification#47 are not part of hotp-verification 1.7 version bump and untested from my side.

Waiting for https://github.com/Nitrokey/nitrokey-hotp-verification/issues/41

Nitrokey shouldn't expect another 40h of work on my side to fix their things. As said in all those PR/issues, if this is Nitrokey decided way to collaborate, I expect Nitrokey to open PR under Heads side with tested code prior of Heads merging things upstream, not the other way around anymore without a legally bounding contract with proper seperation of duties and legal responsabilities.

**Comment by tlaurion** [comment]
Pending fixes to be pushed under Heads by Nitrokey team forward https://github.com/Nitrokey/nitrokey-hotp-verification/issues/52#issue-2743143949

**Comment by tlaurion** [comment]
Physical presence requirement under nk3 firmware will be done later on and is under nitrokey responsibility to provide code changes under Heads in the future.

Comment left at https://github.com/Nitrokey/nitrokey-hotp-verification/issues/41#issuecomment-2555787304

Which is copy of previous comment.
